RFD 0153 Resource Guidelines (#34103)
gravitational/teleport: rfd/0153-resource-guidelines.md

them, see [#29234](https://github.com/gravitational/teleport/issues/29234). In addition to reducing confusion and
fostering a better developer experience, this will also remove complexity from the terraform provider and teleport
operator that have to accommodate for every subtle API and resource difference that exist today.

## Details

### Defining a resource

A resource at minimum MUST include a kind, version, `teleport.header.v1.Metadata` and a specification message with any
additional parameters needed to represent the resource. The kind, version and metadata helps to identify the type of
resource, when it should expire, the version of the resource, etc. This is particularly important for ensuring that all
resources contain the same shape and that tools like `tctl` are able to leverage `services.UnknownResource` to identify
the resource and act appropriately.

```protobuf
import "teleport/header/v1/resourceheader.proto";

// Foo is a resource that does foo.
message Foo {
  // The kind of resource represented.
  string kind = 1;
  // An optional subkind to differentiate variations of the same kind.
  string sub_kind = 2;
  // The version of the resource being represented.

>   string version = 3;

Reviewer: Maybe note somewhere that `CheckAndSetDefaults` should fail on unknown version values, even for the initial release of the resource (its been a common error in the past to not start caring about `Version` until we need to bump it... then having to deal with older clients silently accepting unrecognized versions and doing the wrong thing with them).

Reviewer 2: I think that's fine, thanks.

#### Create

The `Create` RPC takes a resource to be created and must also return the newly created resource so that any fields that
are populated server side are provided to clients without requiring an additional call to `Get`.

The request MUST fail and return a `trace.AlreadyExists` error if a matching resource is already present in the backend.

```protobuf
// Creates a new Foo resource in the backend.
    rpc CreateFoo(CreateFooRequest) returns (CreateFooResponse);

message CreateFooRequest {
  // The desired Foo to be created.
  Foo foo = 1;
}

message CreateFooResponse {
  // The created Foo as it exists in the backend. The only differences between the returned
  // resource and the requested resource are due to properties that are updated by the server.
  // In most cases this will be limited to the Revision field of the Metadata in the ResourceHeader.
  Foo foo = 1;
}
```

When creating a new resource, the `backend.Backend.Create` method should be used to persist the resource. It is also
imperative that the revision generated by the backend is set on the returned resource.

```go
func (s *FooService) CreateFoo(ctx context.Context, foo foov1.Foo) (foov1.Foo, error) {

> value, err := convertFooToValue(foo)

Reviewer: IMO we shouldn't be recommended these custom marshal/unmarshal functions.  Tying behavior to encoding is bad (and makes it a nightmare to handle any resource in a generic manner).

I think that we should recommend using marshal/unmarshal helpers (e.g. `utils.FastMarshal`) for encoding/decoding, and methods (e.g. `CheckAndSetDefaults`) for behavior/validation.  Case in point, all the backend-related example code in this RFD could be written as a generic helper over `T Resource` if we just got rid of the `convert*` functions.

allows users to request the sensitive information. By default, the secrets should not be provided in the response.

```protobuf
// Returns a page of Foo and the token to find the next page of items.
    rpc ListFoo(ListFooRequest) returns (ListFooResponse);

message ListFoosRequest {
  // The maximum number of items to return.
  // The server may impose a different page size at its discretion.
  int32 page_size = 1;
  // The next_page_token value returned from a previous List request, if any.
  string page_token = 2;
  // Specifies whether to load associated secrets with the resource.
  // NOTE: This is only required for resources which have secrets(passwords, 
  // private keys, mfa devices, etc). If a resource does not have any 
  // secrets this can be omitted.
  bool with_secrets = 3;
}

message ListFoosResponse {
  // The page of Foo that matched the request.
  repeated Foo foos = 1;
  // Token to retrieve the next page of results, or empty if there are no
  // more results in the list.
  string next_page_token = 2;
}
```

A listing operation should not abort entirely if a single item cannot be (un)marshalled, it should instead be logged,

> and the rest of the page should be processed. Aborting an entire page when a single entry is invalid causes the cache
> to be permanently unhealthy since it is never able to initialize loading the affected resource.

Reviewer: ```suggestion
and the rest of the page should be processed. Aborting an entire page when a single entry is invalid causes teleport
to be permanently unhealthy since it is never able to load or cache the affected resource.
```

Cache init is probably the first operation that will show up as broken in logs, but all features that might directly or indirectly rely on consuming the resource set will be permanently borked.

them, see [#29234](https://github.com/gravitational/teleport/issues/29234). In addition to reducing confusion and
fostering a better developer experience, this will also remove complexity from the terraform provider and teleport
operator that have to accommodate for every subtle API and resource difference that exist today.

## Details

### Defining a resource

A resource at minimum MUST include a kind, version, `teleport.header.v1.Metadata` and a specification message with any
additional parameters needed to represent the resource. While the kind and version may seem like they would be easy to
derive from the message definition itself, they need to be defined so that anything processing a generic resource can
identify which resource is being processed. For example, `tctl` interacts with resources in their in raw yaml text form
and leverage `services.UnknownResource` to identify the resource and act appropriately.

```protobuf
import "teleport/header/v1/resourceheader.proto";

// Foo is a resource that does foo.
message Foo {
  // The kind of resource represented.
  string kind = 1;
  // An optional subkind to differentiate variations of the same kind.
  string sub_kind = 2;

>   // The version of the resource being represented.

Reviewer 2: It's a good question. I do think an api_version can be useful, as it can be a indicator that certain features are supported by the backend, but I also suspect the main reason it exists in projects like k8s is because YAML has a harder time representing that. We've relied on versions before to toggle different behavior (think roles); whether that's a good idea is debatable, but it adds design flexibility.

I think we can "solve" some of my gripes with documentation, stating clearly what "version" and what "revision" mean, and how one is not the other (as in many other places "version" is used for what we call "revision"). (I would prefer to call it api_version and be done with it, though, but it doesn't have to happen here.)

>   // The version of the resource being represented.

Reviewer 2: Also adding to this, I'd rather never see FooV2 _or_ foo.v2 - API major bumps are a pain and should be a very rare exception to the rule.

  string version = 3;
  // Common metadata that all resources shared.
  teleport.header.v1.Metadata metadata = 4;
  // The resource specific specification.
  FooSpec spec = 5;
}

// FooSpec contains resource specific properties.

> message FooSpec {

Reviewer 2: Why is FooSpec a thing? Why not have this directly on Foo?

Reviewer 2: Fair enough. I couldn't resist the comment.

migrate away from it as described in [RFD-0139](https://github.com/gravitational/teleport/pull/28386).
The `teleport.header.v1.Metadata` is a clone of `types.Metadata` which doesn't use any of the gogoproto features.
Legacy resources also had a `types.ResourceHeader` that used gogo magic to embed the type in the resource message. To
get around this, the required fields from the header MUST be included in the message itself. A non-gogo clone does exist
`teleport.header.v1.ResourceHeader`, however, to get the fields, embeded custom marshalling must be manually written.

If a resource has associated secrets (password, private key, jwt, mfa device, etc.) they should be defined in a separate
resource and stored in a separate key range in the backend. The traditional pattern of defining secrets inline and only
returning them if a `with_sercrets` flag is provided cause a variety of problems and introduce opportunity for human
error to accidentally include secrets when they should not have been. It would then be the responsibility of the caller
to get both the base resource and the corresponding secret resource if required.

> 

Reviewer 2: Bump? I think proto->json conversion and whether we use protos as-is for tctl and the like are still somewhat open.

For example, device trust defines "presentation" protos for tctl, but I'm not sure I would do that again today.

> 

Reviewer 2: > If a resource includes an enum but the desire is for tctl to display a friendly string representation instead of a number then I'd advise to use a presentation type

That was my main motivation, but today I would even question that. The reality is that we have plenty of types that expose numerical enums already, so the work felt like a drop in the ocean (but nevertheless took time and effort).

WDYT? Is this an area we should delve into? (Maybe not now but in a future update of the RFD?)

### API

All APIs should follow the following conventions that are largely based on
the [Google API style guide](https://cloud.google.com/apis/design/standard_methods).

#### Create

The `Create` RPC takes a resource to be created and must also return the newly created resource so that any fields that
are populated server side are provided to clients without requiring an additional call to `Get`.

The request MUST fail and return a `trace.AlreadyExists` error if a matching resource is already present in the backend.

```protobuf
// Creates a new Foo resource in the backend.

>     rpc CreateFoo(CreateFooRequest) returns (CreateFooResponse);

Reviewer 3: I'm open to/supportive of the adoption of "transient" fields within resources as a replacement of returning a wrapped resource - as long as we have some way of returning additional information.

>     rpc CreateFoo(CreateFooRequest) returns (CreateFooResponse);

Reviewer 2: It seems like we are going the direction of transient fields inside "status", as per other thread. I do tend to favor Google style, as it makes it harder for single-entity RPCs like Create/Update/etc to get too creative with their replies.

As for output_fields I'm not sure, as I've never used it before, but I assume it serves primarily as a form of documentation. (Server-side code still needs to be coded appropriately.)

  Foo foo = 1;
}
```

#### Update

The `Update` RPC takes a resource to be updated and must also return the updated resource so that any fields that are
populated server side are provided to clients without requiring an additional call to `Get`. If partial updates of a
resource are desired, the request may contain
a [FieldMask](https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#fieldmask).

The request MUST fail and return a `trace.NotFound` error if there is no matching resource in the backend.

```protobuf
// Updates an existing Foo in the backend.
    rpc UpdateFoo(UpdateFooRequest) returns (UpdateFooResponse);

message UpdateFooRequest {
  // The full Foo resource to update in the backend.
  Foo foo = 1;
  // A partial update for an existing Foo resource.
  FieldMask update_mask = 2;
}

message UpdateFooResponse {
  // The updated Foo as it exists in the backend. The only differences between the returned
  // resource and the requested resource are due to properties that are updated by the server.
  // In most cases this will be limited to the Revision field of the Metadata in the ResourceHeader.
  Foo foo = 1;

> }

Reviewer 2: Fair enough. We could document explicitly what Metadata fields are expected on updates, but a good error message also goes a long way here.

```

> #### Upsert

Reviewer 2: > It's also worth noting that tctl create and even operator don't work well with the idea of system-managed fields (...)

This is now solved by the status field.

    rpc GetFoo(GetFooRequest) returns (GetFooResponse);

message GetFooRequest {
  // A filter to match the Foo by. Some resource may require more parameters to match and
  // may not use the name at all.
  string name = 1;
}

message GetFooResponse {
  // The matched Foo resource.
  Foo foo = 1;
}
```

#### List

The `List` RPC takes the requested page size and starting point and returns a list of resources that match. If there are
additional resources, the response MUST also include a token that indicates where the next page of results begins.

Most legacy APIs do not provide a paginated way to retrieve resources and instead offer some kind of `GetAllFoos` RPC
which either returns all `Foo` in a single message or leverages a server side stream to send each `Foo` one at a time.
Returning all items in a single message causes problems when the number of resources scales beyond gRPC message size
limits. Due to caches making use of the `GetAll` RPC variants to populate themselves during initialization this can lead
to permanently unhealthy caches constantly retrying to initialize which can lead to backend throttling and enough
increased load on Auth to render the cluster unusable.

> 

Reviewer 4: I think it's worth adding a recommended way for doing a filter operation.
Even if it's just an optional section.

> 

Reviewer 4: I was thinking about defining what is a good well-typed filter.
Something like:
To, optionally, implement a search functionality you must define a `ListWithFilters` method which receives a page_size, page_token and a OneOf ListFoosFilter.
This last element describes the filter you want to implement
Eg
```
message FilterFoosSystemBarAndQux {}
```

Or something else totally different.

We could go the `key:values` that AWS APIs have
https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeInstances.html

Or the `key:operator:values` that Kube has
https://kubernetes.io/docs/concepts/overview/working-with-objects/field-selectors/

But those are not well-typed
---
As for a real need, I've been thinking about adding a ListWithFilter to a particular resource.
Get all the DiscoveryConfigs that have a given DiscoveryGroup (a string within the resource's `spec`)

> 

Reviewer 2: I still think list filters are better considered case-by-case, with rich/strong-typed filters as the preference. I wouldn't be comfortable making a call for all filtered-List methods at this moment, nor do I think it's the time to adopt generic alternatives like AWS or K8s.

I was looking for examples of my own, but the best I've found is [FindDevices](https://github.com/gravitational/teleport/blob/eadb331982a10455abd0e4a80876d7048748a174/api/proto/teleport/devicetrust/v1/devicetrust_service.proto#L90), which is custom tailored for `tctl get devices/X` and similar operations. It's not a complex search, but it somewhat illustrates the simplicity point.

> and cannot be expressed via a query to the actual backend

I don't think this is a great idea - suddenly our backend schema becomes tied to the public API and can't be changed. I think this kind of generic filter is better expressed over a "virtual" (but well-described) table or over the type itself (like RSQL). This way we don't expose storage and retain the ability to implement the actual queries as we find it best. Still, this is typically a weak-type system that requires more logic than more specific filters.

> 

Reviewer 4: I agree with having this defined on a case-by-case :+1:
However, giving one example of how those can be built, is IMO worth.
Otherwise, we'll end up with very different patterns.

}

message ListFoosResponse {
  // The page of Foo that matched the request.
  repeated Foo foos = 1;
  // Token to retrieve the next page of results, or empty if there are no
  // more results in the list.
  string next_page_token = 2;
}
```

A listing operation should not abort entirely if a single item cannot be (un)marshalled, it should instead be logged,
and the rest of the page should be processed. Aborting an entire page when a single entry is invalid causes the cache
to be permanently unhealthy since it is never able to initialize loading the affected resource.

#### Delete

The `Delete` RPC takes the parameters required to match a resource (usually the resource name should suffice), and
removes the specified resource from the backend and returns a `google.protobuf.Empty`.

The request MUST fail and return a `trace.NotFound` error if there is no matching resource in the backend.

```protobuf
// Remove a matching Foo resource
    rpc DeleteFoo(DeleteFooRequest) returns (google.protobuf.Empty);

message DeleteFooRequest {
  // Name of the foo to remove. Some resource may require more parameters to match and
  // may not use the name at all.
  string name = 1;
}
```

> 

Reviewer 2: Looks good, thanks.

### Backend Storage

A backend service to handle persisting and retrieving a resource from the backend is typically defined in

> `lib/services/local/foo.go`. An accompanying interface which mirrors the service is defined in `lib/services/foo.go`.

Reviewer 3: I agree with Alan here - anything we can do to move away from the "omnipackages" that cover almost all Teleport domains is good and makes the codebase much more manageable to work on for Teams that only touch a small number of the domains (as well as having potential benefits in terms of binary size etc) - but I wonder if this falls out of scope of this RFD ?

> `lib/services/local/foo.go`. An accompanying interface which mirrors the service is defined in `lib/services/foo.go`.

Reviewer 2: I understand. I'd be happy not recommending lib/services/local, if we can get away with that.

Continuing on with the example above, the backend service for the `Foo` resource might look like the following.

#### Create

When creating a new resource, the `backend.Backend.Create` method should be used to persist the resource. It is also
imperative that the revision generated by the backend is set on the returned resource.

```go
func (s *FooService) CreateFoo(ctx context.Context, foo *foov1.Foo) (*foov1.Foo, error) {
    value, err := convertFooToValue(foo)
    if err != nil {
        return nil, trace.Wrap(err)
    }
    item := backend.Item{
        Key:     backend.Key("foo", foo.GetName()),
        Value:   value,
        Expires: foo.Expiry(),
    }

    lease, err := s.backend.Create(ctx, item)
    if err != nil {
        return nil, trace.Wrap(err)
    }

    // Update the foo with the revision generated by the backend during the write operation.

>     foo.SetRevision(lease.Revision)

Reviewer 2: How's the SetRevision method getting into the resource? Should this read `foo.GetMetadata().SetRevision(lease.Revision)` ?

>     foo.SetRevision(lease.Revision)

Reviewer 2: Yes, but this means we'll keep adding methods to generated types, a definite smell. It wouldn't play well with gen/ packages either, as we delete those during codegen (not by accident 😉).

for an example.

```go
func (s *FooService) UpsertFoo(ctx context.Context, foo *foov1.Foo) (*foov1.Foo, error) {
    value, err := convertFooToValue(foo)
    if err != nil {
        return nil, trace.Wrap(err)
    }
    item := backend.Item{
        Key:     backend.Key("foo", foo.GetName()),
        Value:   value,
        Expires: foo.Expiry(),
    }

    lease, err := s.backend.Put(ctx, item)
    if err != nil {
        return nil, trace.Wrap(err)
    }

    // Update the foo with the revision generated by the backend during the write operation.
    foo.SetRevision(lease.Revision)
    return foo, nil
}
```

#### Get

To retrieve a resource the `backend.Backend.Get` method should be provided a key built from the match parameters of the
request.

```go
func (s *AccessService) GetFoo(ctx context.Context, name string) (*Foo, error) {
    if name == "" {
        return nil, trace.BadParameter("missing foo name")
    }

    item, err := s.backend.Get(ctx, backend.Key("foo", name))
    if err != nil {
        if trace.IsNotFound(err) {

>             return nil, trace.NotFound("foo %v is not found", name)

Reviewer 2: I see. Could we push that into the backend, though? It seems laborious having this everywhere, plus I'm sure I'll forget it (and leak the key like you said).

>             return nil, trace.NotFound("foo %v is not found", name)

Reviewer 2: I was going to suggest that it could do an educated guess based on the last part of the key, but that isn't necessarily correct or friendly. Oh, well.

            if err != nil {
                s.log.Warnf("Skipping foo at %s because conversion from backend item failed: %v", item.Key, err)
                return nil, true
            }
            return foo, true
        })

    foos, more := stream.Take(userStream, pageSize)
    var nextToken string
    if more && fooStream.Next() {    
        nextToken = backend.NextPaginationKey(foos[len(foos)-1])
    }

    return foos, nextToken, trace.NewAggregate(err, fooStream.Done())
}
```

### Cache

One thing to consider when creating a resource is whether it will need to be cached. As mentioned above, any resource
that is cached must have its backend layer implement the specific set of operations required by the cache collections
[executor](https://github.com/gravitational/teleport/blob/004d0db0c1f6e9b312d0b0e1330b6e5bf1ffef6e/lib/cache/collections.go#L54-L76).
To avoid including `Upsert` or `DeleteAll` in the gRPC API if they are only consumed by the cache is to define a local
variant of the backend service similar to
[`services.DynamicAccessExt`](https://github.com/gravitational/teleport/blob/004d0db0c1f6e9b312d0b0e1330b6e5bf1ffef6e/lib/services/access_request.go#L260-L278).

> Not all resources need to be cached. If a resource is infrequently accessed outside the hot path, then adding it to the

Reviewer: Caching is most important when a resource is accessed frequently and in a "hot path" (i.e. during the process of performing normal day to day operations).  The best examples of 'must cache' items are things that are read during basic teleport function/connectivity (e.g. cluster networking config, CAs, etc), and resources that need to be loaded during the process of finding and accessing services (e.g. nodes, databases, roles, etc).  All of these may often be loaded tens or hundreds of times during a user's day to day usage of teleport.

> Not all resources need to be cached. If a resource is infrequently accessed outside the hot path, then adding it to the

Reviewer 2: Thanks, this is great advice. @rosstimothy, could we add Forrest guidelines to the design too?

    return cache.Foo.DeleteFoo(ctx, &foov1.DeleteFoo{Name: resource.GetName()})
}
```

### Proto Specification

Below is the entire specification for the examples above.

<details open><summary>Foo Proto</summary>

```protobuf
syntax = "proto3";

package teleport.foo.v1;

import "teleport/header/v1/resourceheader.proto";

option go_package = "github.com/gravitational/teleport/api/gen/proto/go/teleport/foo/v1;foov1";

// Foo is a resource that does foo.
message Foo {
  // The kind of resource represented.
  string kind = 1;
  // An optional subkind to differentiate variations of the same kind.
  string sub_kind = 2;
  // The version of the resource being represented.
  string version = 3;
  // Common metadata that all resources shared.
  teleport.header.v1.Metadata metadata = 4;
  // The resource specific specification.
  FooSpec spec = 5;
}

// FooSpec contains resource specific properties.
message FooSpec {
  string bar = 1;
  int32 baz = 2;
  bool qux = 3;
}

// FooService provides an API to manage Foos.
service FooService {
  // GetFoo returns the specified Foo resource.
  rpc GetFoo(GetFooRequest) returns (GetFooResponse);

>   // ListFoo returns a page of Foo resources.
>   rpc ListFoo(ListFooRequest) returns (ListFooResponse);

Reviewer 2: nit:

```suggestion
  // ListFoos returns a page of Foo resources.
  rpc ListFoos(ListFoosRequest) returns (ListFoosResponse);
```

  // CreateFoo creates a new Foo resource.
  rpc CreateFoo(CreateFooRequest) returns (CreateFooResponse);

  // UpdateFoo updates an existing Foo resource.
  rpc UpdateFoo(UpdateFooRequest) returns (UpdateFooResponse);

  // UpsertFoo creates or updates a Foo resource.
  rpc UpsertFoo(UpsertFooRequest) returns (UpsertFooResponse);

>   // DeleteFoo removes the specified Foo resource.

Reviewer 2: ```suggestion
  // DeleteFoo hard-deletes the specified Foo resource.
```

Maybe?

  rpc DeleteFoo(DeleteFooRequest) returns (google.protobuf.Empty);
}

// Request for GetFoo.
message GetFooRequest {
  // The name of the Foo resource to retrieve.

>   string name = 1;

Reviewer 2: Suggestion:

```suggestion
  string foo_id = 1;
```

Same for similar fields.

The strictest validation of a resource should be performed prior to write operations. Any resource persisted in the
backend should be guaranteed to be valid. Read operations should not perform resource validations, doing so could
prevent a resource being read if validations are modified to be more restrictive after a resource had already been
written. This allows for resilient reads which ensure that any resource stored is always allowed to be retrieved from
the backend.

#### Create

When creating a new resource, the `backend.Backend.Create` method should be used to persist the resource. It is also
imperative that the revision generated by the backend is set on the returned resource.

```go
func (s *FooService) CreateFoo(ctx context.Context, foo *foov1.Foo) (*foov1.Foo, error) {
	value, err := convertFooToValue(foo)
	if err != nil {
		return nil, trace.Wrap(err)
	}
	item := backend.Item{
		Key:     backend.Key("foo", foo.GetName()),
		Value:   value,
		Expires: foo.Expiry(),
	}

	lease, err := s.backend.Create(ctx, item)
	if err != nil {
		return nil, trace.Wrap(err)
	}

	// Update the foo with the revision generated by the backend during the write operation.

> 	foo.GetMetadata().SetRevision(lease.Revision)

Reviewer 5: This isn't done in `CreateResource` in `service.Generic` on master: https://github.com/gravitational/teleport/blob/1dcd7b37a673eb813245ea041a18e934f6061bea/lib/services/local/generic/generic.go#L192C4-L205

```go
// CreateResource creates a new resource.
func (s *Service[T]) CreateResource(ctx context.Context, resource T) error {
        item, err := s.MakeBackendItem(resource, resource.GetName())
        if err != nil {
                return trace.Wrap(err)
        }

        _, err = s.backend.Create(ctx, item)
        if trace.IsAlreadyExists(err) {
                return trace.AlreadyExists("%s %q already exists", s.resourceKind, resource.GetName())
        }

        return trace.Wrap(err)
}
```

Should it be updated?

Use the non-exhaustive list below as a guide when adding a new resource. Further sections in the RFD dive into more
detail about what is needed to complete a particular item.

- [ ] Create proto spec for resource and RPC service
- [ ] Create backend service
- [ ] Add resource client to `api` client
- [ ] Implement gRPC service
- [ ] Add resource support to tctl (get, create, edit)
- [ ] Optional: Add resource to cache
- [ ] Add support for bootstrapping the resource
- [ ] Add support for resource to Teleport Operator
- [ ] Add support for resource to Teleport Terraform Provider

### Defining a resource

A resource at minimum MUST include a kind, version, `teleport.header.v1.Metadata` and a specification message with any
additional parameters needed to represent the resource. While the kind and version may seem like they would be easy to
derive from the message definition itself, they need to be defined so that anything processing a generic resource can
identify which resource is being processed. For example, `tctl` interacts with resources in their in raw yaml text form
and leverages `services.UnknownResource` to identify the resource and act appropriately.

```protobuf

> import "teleport/header/v1/resourceheader.proto";

Reviewer 5: ```suggestion
import "teleport/header/v1/metadata.proto";
```

`tctl create`, then any fields within the `spec` MUST not be altered dynamically
by the Teleport process. When Teleport automatically modifies the `spec` during
runtime it causes drift between external IaC. If a resource has properties that
are required to be modified dynamically by Teleport, a separate `status` field
should be added to the resource to contain them. These fields will be ignored
by IaC tools during their reconcilliation.

### API

All APIs should follow the conventions listed below that are largely based on
the [Google API style guide](https://cloud.google.com/apis/design/standard_methods).

#### Create

The `Create` RPC takes a resource to be created and must also return the newly created resource so that any fields that
are populated server side are provided to clients without requiring an additional call to `Get`.

The request MUST fail and return a `trace.AlreadyExists` error if a matching resource is already present in the backend.

```protobuf
// Creates a new Foo resource in the backend.
    rpc CreateFoo(CreateFooRequest) returns (Foo);

message CreateFooRequest {
  // The desired Foo to be created.
  Foo foo = 1;
}
```

> #### Update

Reviewer 6: There's some discussion of update/upsert over in https://github.com/gravitational/teleport/pull/35441#discussion_r1428614490, I will suggest we move that here.

Reviewer 6: What about my point about the uselessness of patch-style operations under this paradigm (which pertains to your `FieldMask` comment below). As I understand it, in order to do an `UpdateFoo` with optimistic locking, you will need to have first done a `GetFoo` to get the object's revision for the call to `Update`. So then wouldn't that make the `FieldMask` useless, since you'll have the entire object on-hand anyhow?

Reviewer 6: Got it, yeah that use case makes sense.

That brings up a question for me, though: do we need to implement a patch-style fieldmask operation in each supported `Backend` individually. Or is there some existing abstraction that takes care of that for us? I don't know that I've ever actually added a new resource to Teleport, so I'm not quite sure what that process entails generally, nor for a resource that supports FieldMask specifically.

Reviewer 6: Thanks, that's similar although I think I'm also asking a further question -- as I understand it we support various backends like sqlite, etcd, dynamodb, etc. In adding a new resource, would we also need to understand how patching works in each of those and extend them?

Regardless of the answer, since this is likely to be used as a reference doc it might be worthwhile including a link to somewhere in the codebase where field masks were done well.

Reviewer 3: The new BotService (https://github.com/gravitational/teleport/pull/35441) also makes use of field_masks for patching.

flag: `teleport start --bootstrap=/some/path/to/resources/yaml`.

For a resource to be supported it must be added to the list of items retrieved with `tctl get all` and to the auth
initialization code responsible for parsing resources during
the [bootstrap process](https://github.com/gravitational/teleport/blob/d0f2b4406bfacc895f796b665d07c5d740280e38/lib/auth/init.go#L321-L335).

### Backward Compatibility

Changing existing resources which do not follow the guidelines laid out in this RFD may lead to breaking changes. It is
not recommended to change existing resources for change’s sake. Migrating APIs which do not conform to the
recommendations in this RFD can be made in a backward compatible manner. This can be achieved by adding new APIs that
conform with the advice above and falling back to the existing APIs if a `trace.NotImplemented` error is received. Once
all compatible versions of Teleport are using the new version of the API, the old API may be cleaned up.

### Proto Specification

Below is the entire specification for the examples above.

<details open><summary>Foo Proto</summary>

```protobuf
syntax = "proto3";

package teleport.foo.v1;

> import "teleport/header/v1/resourceheader.proto";

Reviewer 5: ```suggestion
import "teleport/header/v1/metadata.proto";
```

  // the creator/owner of the resource and not dynamically altered or updated.
  FooSpec spec = 5;
  // Any dynamic state of Foo that is modified during runtime of the
  // Teleport process.
  FooStatus status = 6;
}

// FooSpec contains specific properties of a Foo that MUST only
// be modifed by the owner of the resource. These properties should
// not be auotmatically adjusted by Teleport during runtime.
message FooSpec {
  string bar = 1;
  int32 baz = 2;
  bool qux = 3;
}

// FooStatus contains dynamic properties of a Foo. These properties are
// modified during runtime of a Teleport process. They should not be exposed
// to end users and ignored by external IaC tools.
message FooStatus {
  google.protobuf.Timestamp next_audit = 1;
  string teleport_host = 2;
}
```

</details>

<details open><summary>Foo Service</summary>

```protobuf
syntax = "proto3";

package teleport.foo.v1;

import "teleport/foo/v1/foo.proto";

option go_package = "github.com/gravitational/teleport/api/gen/proto/go/teleport/foo/v1;foov1";

// FooService provides an API to manage Foos.
service FooService {
  // GetFoo returns the specified Foo resource.

>   rpc GetFoo(GetFooRequest) returns (GetFooResponse);

Reviewer 2: ```suggestion
  rpc GetFoo(GetFooRequest) returns (Foo);
```

  // ListFoos returns a page of Foo resources.
  rpc ListFoos(ListFoosRequest) returns (ListFoosResponse);

  // CreateFoo creates a new Foo resource.
  rpc CreateFoo(CreateFooRequest) returns (Foo);

  // UpdateFoo updates an existing Foo resource.
  rpc UpdateFoo(UpdateFooRequest) returns (Foo);

  // UpsertFoo creates or replaces a Foo resource.
  rpc UpsertFoo(UpsertFooRequest) returns (Foo);

  // DeleteFoo hard deletes the specified Foo resource.
  rpc DeleteFoo(DeleteFooRequest) returns (google.protobuf.Empty);
}

// Request for GetFoo.
message GetFooRequest {
  // The id of the Foo resource to retrieve.
  string foo_id = 1;
}

> 
> // Response for GetFoo.
> message GetFooResponse {
>   // The foo matching the request filters.
>   Foo foo = 1;
> }

Reviewer 2: ```suggestion
```

- [ ] Optional: Add support for resource to Teleport Terraform Provider

### Defining a resource

A resource at minimum MUST include a kind, version, `teleport.header.v1.Metadata` and a specification message with any
additional parameters needed to represent the resource. While the kind and version may seem like they would be easy to
derive from the message definition itself, they need to be defined so that anything processing a generic resource can
identify which resource is being processed. For example, `tctl` interacts with resources in their in raw yaml text form
and leverages `services.UnknownResource` to identify the resource and act appropriately.

```protobuf
import "teleport/header/v1/resourceheader.proto";

// Foo is a resource that does foo.
message Foo {
  // The kind of resource represented.
  string kind = 1;
  // Differentiates variations of the same kind. All resources should
  // contain one, even if it is never populated.
  string sub_kind = 2;
  // The version of the resource being represented.
  string version = 3;
  // Common metadata that all resources share.
  teleport.header.v1.Metadata metadata = 4;
  // The specific properties of a Foo.

>   FooSpec spec = 5;

Reviewer 3: From a Machine ID perspective, I'd certainly much prefer a typed `status` field to store things like the cert revision in rather than the current solution using the metadata.

Generally, I like the idea of a `status` field, especially in terms of how we interact with IaC and it allows us to clearly mark which fields should not be modified by users.

I suppose an alternative option is to just store these sorts of things in another resource - but this adds a lot of complexity.

>   FooSpec spec = 5;

Reviewer 2: Are we OK having `spec` fields that are solely managed by Teleport, or should we consider these a part of `status` too?

>   FooSpec spec = 5;

Reviewer 7: In my opinion, we have two cases:
- the resource is created by the user/exposed to the user -> everything in spec should be managed by the user, if something is managed by teleport, it should be out of spec
- the resource is not created by the user/exposed via `tctl` -> we do what we want

If we have fields in the spec that are managed by teleport and some managed by the user, it means we'll have to manually tell each IaC tool to skip reconciling those fields (and we'll forget some fields and end up with something broken).

>   FooSpec spec = 5;

Reviewer 4: I think the Database resource is a good example of this:
There's a `spec` and a `status`:
https://github.com/gravitational/teleport/blob/464edfa2cdc7acc4e8d2b711fef3e9740e21d4d5/api/proto/teleport/legacy/types/types.proto#L260-L268

>   FooSpec spec = 5;

Reviewer 3: LGTM!

// Creates a new Foo or replaces an existing Foo in the backend.
    rpc UpsertFoo(UpsertFooRequest) returns (UpsertFooResponse);

message UpsertFooRequest {
  // The full Foo resource to persist in the backend.
  Foo foo = 1;
}

message UpsertFooResponse {
  // The upserted Foo as it exists in the backend. The only differences between the returned
  // resource and the requested resource are due to properties that are updated by the server.
  // In most cases this will be limited to the Revision field of the Metadata in the ResourceHeader.
  Foo foo = 1;
}
```

#### Get

The `Get` RPC takes the parameters required to match a resource (usually the resource name should suffice), and returns
the matched resource.

The request MUST fail and return a `trace.NotFound` error if there is no matching resource in the backend.

```protobuf
// Returns a single Foo matching the request
    rpc GetFoo(GetFooRequest) returns (GetFooResponse);

message GetFooRequest {
  // A filter to match the Foo by. Some resource may require more parameters to match and
  // may not use the name at all.
  string foo_id = 1;
}

message GetFooResponse {
  // The matched Foo resource.
  Foo foo = 1;
}
```

> #### List

Reviewer 4: I think we should add a note regarding the expected ordering.
Resource's `name` field ascending, perhaps.

Even if we add a disclaimer saying that clients must not depend on the ordering.

Reviewer 2: If there's pagination then there must be some order. I think the order guarantees should be specific by each RPC, but it's a good idea to encourage APIs to make explicit decisions (as in, say whether there is a guaranteed order or not).

---
authors: Tim Ross (tim.ross@goteleport.com)
state: draft
---

# RFD 153 - Resource Guidelines

## Required approvers

- Engineering: `@zmb3`

## What

Guidelines for creating backend resources and APIs to interact with them.

## Why

To date resources follow slightly different patterns and have inconsistent APIs for creating, updating and deleting
them, see [#29234](https://github.com/gravitational/teleport/issues/29234). In addition to reducing confusion and
fostering a better developer experience, this will also remove complexity from the terraform provider and teleport
operator that have to accommodate for every subtle API and resource difference that exist today.

## Details

### Project structure

All protos are defined in the api module under the proto directory. When adding a new resource and gRPC API a new folder

> that matches the desired package name of the proto, generally a friendly name of the resource, should be created. Inside

Reviewer 3: Nit: In my head, I quite like the word "domain" to describe the appropriate groupings - we probably don't necessarily expect a package per resource where several resources are tightly coupled/related.

#### Create

The `Create` RPC takes a resource to be created and must also return the newly created resource so that any fields that
are populated server side are provided to clients without requiring an additional call to `Get`.

The request MUST fail and return a `trace.AlreadyExists` error if a matching resource is already present in the backend.

```protobuf
// Creates a new Foo resource in the backend.
    rpc CreateFoo(CreateFooRequest) returns (CreateFooResponse);

message CreateFooRequest {
  // The desired Foo to be created.
  Foo foo = 1;
}

message CreateFooResponse {
  // The created Foo as it exists in the backend. The only differences between the returned
  // resource and the requested resource are due to properties that are updated by the server.
  // In most cases this will be limited to the Revision field of the Metadata in the ResourceHeader.
  Foo foo = 1;
}
```

#### Update

The `Update` RPC takes a resource to be updated and must also return the updated resource so that any fields that are
populated server side are provided to clients without requiring an additional call to `Get`. If partial updates of a
resource are desired, the request may contain

> a [FieldMask](https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#fieldmask).

Reviewer 8: ```suggestion
a [FieldMask](https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#field-mask).
```

> a [FieldMask](https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#fieldmask).

Reviewer 8: Are field masks handled solely by the gRPC package or does the server needs to call some specific functions from the gRPC package to accommodate for field masks? If the latter, it'd be nice to have a link to an example server-side implementation.

Reviewer 8: Ah, I see. I don't know why I thought that maybe gRPC provides a way to "automatically" transform field masks into actual field updates so that I don't have to write a function like `applyDeviceUpdateMask` – that wouldn't make much sense. 🤔

Reviewer 2: It's a common question. If you dig you'll likely find some reflect trickery, but usually I just recommend people to write the for/switch. It's simple, "dumb" code.

}
```

A listing operation should not abort entirely if a single item cannot be (un)marshalled, it should instead be logged,
and the rest of the page should be processed. Aborting an entire page when a single entry is invalid causes the cache
to be permanently unhealthy since it is never able to initialize loading the affected resource.

#### Delete

The `Delete` RPC takes the parameters required to match a resource and performs a hard delete of the specified resource
from the backend and returns a `google.protobuf.Empty`.

The request MUST fail and return a `trace.NotFound` error if there is no matching resource in the backend.

```protobuf
// Remove a matching Foo resource
    rpc DeleteFoo(DeleteFooRequest) returns (google.protobuf.Empty);

message DeleteFooRequest {
  // Name of the foo to remove. Some resource may require more parameters to match and
  // may not use the name at all.
  string foo_id = 1;
}
```

### Backend Storage

A backend service to handle persisting and retrieving a resource from the backend is typically defined in
`lib/services/local/foo.go`. An accompanying interface which mirrors the service is defined in `lib/services/foo.go`.

> Continuing on with the example above, the backend service for the `Foo` resource might look like the following.
> 
> The sections below contain a reference example for how to interact with the backend to perform common operations on a

Reviewer 8: ```suggestion
Continuing on with the example above, the sections below show how the backend service for the `Foo` resource might look like.

The sections also contain a reference example for how to interact with the backend to perform common operations on a
```

At the moment the text says "following" but nothing follows the paragraph, making it seem like something was omitted by mistake.

While `Upsert` and `DeleteAll` semantics are required by the cache it is preferred that the two methods are not directly
exposed in the gRPC API. Several existing resources include a `DeleteAll` purely for the cache that always returns a
`trace.NotImplemented` error. To avoid exposing the methods in the gRPC API at all, a local variant of the backend
service similar
to [`services.DynamicAccessExt`](https://github.com/gravitational/teleport/blob/004d0db0c1f6e9b312d0b0e1330b6e5bf1ffef6e/lib/services/access_request.go#L260-L278)
should be used.

Caching is most important for resources that are accessed frequently and in a "hot path" (i.e., during the process of
performing normal day-to-day operations). For example, resources like cluster networking config, session recording
config, CAs, roles, etc. which are retrieved per connection should be cached to reduce latency. Resources which are
accessed infrequently, or which scale linearly with cluster size are good examples of resources that should NOT be
cached.

If a resource is to be cached, it must be added to

> the[Auth cache](https://github.com/gravitational/teleport/blob/004d0db0c1f6e9b312d0b0e1330b6e5bf1ffef6e/lib/cache/cache.go#L95-L154)

Reviewer 8: ```suggestion
the [Auth cache](https://github.com/gravitational/teleport/blob/004d0db0c1f6e9b312d0b0e1330b6e5bf1ffef6e/lib/cache/cache.go#L95-L154)
```

> and the cache of any service that requires it. To add the `Foo` resource to the cached its executor would look similar

Reviewer 8: ```suggestion
and the cache of any service that requires it. To add the `Foo` resource to the cache its executor would look similar
```

}

func (fooExecutor) deleteAll(ctx context.Context, cache *Cache) error {
	return cache.FooLocal.DeleteAllFoos(ctx)
}

func (fooExecutor) delete(ctx context.Context, cache *Cache, resource types.Resource) error {
	return cache.Foo.DeleteFoo(ctx, &foov1.DeleteFoo{Name: resource.GetName()})
}
```

### api client

For users to interact with the new Foo RPC service the client in the `api`
module needs to be updated to provide the functionality. To do so a new method
should be added to `client.Client` that exposes the gRPC client for the service
as shown below.

```go
func (c *Client) FooClient() foopb.FooServiceClient {
 	return foopb.NewFooServiceClient(c.conn)
}
```

### Bootstrap

Teleport allows a fresh cluster to be created with a set of resources via the `--bootstrap` flag. This is primarily used
when creating a new cluster from a backup of another, or migrating an existing cluster from one storage backend to
another. Typically resources are retrieved from an existing cluster
via `tctl get all --with-secrets > /some/path/to/resources/yaml` and then spawning a new instance of with the bootstrap
flag: `teleport start --bootstrap=/some/path/to/resources/yaml`.

> For a resource to be supported it added to the list of items retrieved with `tctl get all` and to the auth

Reviewer 8: It seems like this sentence was edited and some words are missing.

  // An optional subkind to differentiate variations of the same kind.
  string sub_kind = 2;
  // The version of the resource being represented.
  string version = 3;
  // Common metadata that all resources shared.
  teleport.header.v1.Metadata metadata = 4;
  // The resource specific specification.
  FooSpec spec = 5;
}

// FooSpec contains resource specific properties.
message FooSpec {
  string bar = 1;
  int32 baz = 2;
  bool qux = 3;
}
```

</details>

<details open><summary>Foo Service</summary>

```protobuf
syntax = "proto3";

package teleport.foo.v1;

import "teleport/foo/v1/foo.proto";

option go_package = "github.com/gravitational/teleport/api/gen/proto/go/teleport/foo/v1;foov1";

// FooService provides an API to manage Foos.
service FooService {
  // GetFoo returns the specified Foo resource.
  rpc GetFoo(GetFooRequest) returns (GetFooResponse);

  // ListFoos returns a page of Foo resources.
  rpc ListFoos(ListFoosRequest) returns (ListFoosResponse);

  // CreateFoo creates a new Foo resource.
  rpc CreateFoo(CreateFooRequest) returns (CreateFooResponse);

  // UpdateFoo updates an existing Foo resource.
  rpc UpdateFoo(UpdateFooRequest) returns (UpdateFooResponse);

>   // UpsertFoo creates or updates a Foo resource.

Reviewer 3: Nit

```suggestion
  // UpsertFoo creates or replaces a Foo resource.
```

another. Typically resources are retrieved from an existing cluster
via `tctl get all --with-secrets > /some/path/to/resources/yaml` and then spawning a new instance of with the bootstrap
flag: `teleport start --bootstrap=/some/path/to/resources/yaml`.

For a resource to be supported it added to the list of items retrieved with `tctl get all` and to the auth
initialization code responsible for parsing resources during
the [bootstrap process](https://github.com/gravitational/teleport/blob/d0f2b4406bfacc895f796b665d07c5d740280e38/lib/auth/init.go#L321-L335).

### Backward Compatibility

Changing existing resources which do not follow the guidelines laid out in this RFD may lead to breaking changes. It is
not recommended to change existing resources for change’s sake. Migrating APIs which do not conform to the
recommendations in this RFD can be made in a backward compatible manner. This can be achieved by adding new APIs that
conform with the advice above and falling back to the existing APIs if a `trace.NotImplemented` error is received. Once
all compatible versions of Teleport are using the new version of the API, the old API may be cleaned up.

> 
> ### P
> 

Reviewer 2: nit: leftover text?

```suggestion
```

// Foo is a resource that does foo.
message Foo {
  // The kind of resource represented.
  string kind = 1;
  // Differentiates variations of the same kind. All resources should
  // contain one, even if it is never populated.
  string sub_kind = 2;
  // The version of the resource being represented.
  string version = 3;
  // Common metadata that all resources share.
  teleport.header.v1.Metadata metadata = 4;
  // The specific properties of a Foo. These should only be modified by
  // the creator/owner of the resource and not dynamically altered or updated.
  FooSpec spec = 5;
  // Any dynamic state of Foo that is modified during runtime of the
  // Teleport process.
  FooStatus status = 6;
}

// FooSpec contains specific properties of a Foo that MUST only
// be modifed by the owner of the resource. These properties should
// not be auotmatically adjusted by Teleport during runtime.
message FooSpec {
  string bar = 1;
  int32 baz = 2;
  bool qux = 3;
}

// FooStatus contains dynamic properties of a Foo. These properties are
// modified during runtime of a Teleport process. They should not be exposed
// to end users and ignored by external IaC tools.

> message FooSpec {

Reviewer 4: ```suggestion
message FooStatus {
```

├── buf.yaml
└── teleport
    ├── foo
    │   └── v1
    │       ├── foo.proto
    │       └── foo_service.proto
    ├── legacy
    │   ├── client
    │   │   └── proto
    │   │     ├── authservice.proto
    │   │     ├── certs.proto
    │   │     ├── event.proto
    │   │     ├── joinservice.proto
    │   │     └── proxyservice.proto
    │   └── types
    │         ├── device.proto
    │         ├── events
    │         │    ├── athena.proto
    │         │    └── events.proto
    │         ├── types.proto
    │         ├── webauthn
    │         │    └── webauthn.proto
    │         └── wrappers
    │              └── wrappers.proto
```

The legacy directory contains resources and API definitions that were defined prior to our shift to user smaller,
localized services per resource. Adding new resources and APIs to the giant monolithic `proto.AuthService` should be
avoided if possible.

### New resource checklist

Use the non-exhaustive list below as a guide when adding a new resource. Further sections in the RFD dive into more
detail about what is needed to complete a particular item.

- [ ] Create proto spec for resource and RPC service
- [ ] Create backend service

> - [ ] Add resource support to api client

Reviewer 2: Could we avoid that and, instead, add the new gRPC client itself to the API client? It makes the API client smaller and does away with a bunch of boilerplate.

Ie: https://github.com/gravitational/teleport/blob/b65e7f3b7e864a9e0737922948b46ef9e9d8675a/api/client/client.go#L730-L732

Reviewer 2: Gotcha. I think it's worth spelling out, otherwise we might see method duplication there (as was the norm).

- [ ] Implement gRPC service

> - [ ] Add support for admin operations to tctl

Reviewer 2: Are these the resource-like operations? Create, get, etc?

- [ ] Optional: Add resource to cache
- [ ] Optional: Add support for bootstrapping the resource
- [ ] Optional: Add support for resource to Teleport Operator

> - [ ] Optional: Add support for resource to Teleport Terraform Provider

Reviewer 2: Could we add a guideline for when this needs to happen?
